File: src/android/facebook_ads/src/main/java/com/ee/FacebookAdsBridge.kt
package com.ee

import android.app.Activity
import android.app.Application
import androidx.annotation.AnyThread
import com.ee.facebook.ads.BuildConfig
import com.ee.internal.FacebookBannerAd
import com.ee.internal.FacebookBannerHelper
import com.ee.internal.FacebookInterstitialAd
import com.ee.internal.FacebookNativeAd
import com.ee.internal.FacebookRewardedAd
import com.ee.internal.deserialize
import com.ee.internal.serialize
import com.facebook.ads.AdSettings
import com.facebook.ads.AdSize
import com.facebook.ads.AudienceNetworkAds
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.serialization.InternalSerializationApi
import kotlinx.serialization.Serializable
import java.util.concurrent.ConcurrentHashMap
import kotlin.coroutines.resume

/**
 * Created by Pham Xuan Han on 17/05/17.
 */
@InternalSerializationApi
class FacebookAdsBridge(
    private val _bridge: IMessageBridge,
    private val _logger: ILogger,
    private val _application: Application,
    private var _activity: Activity?) : IPlugin {
    companion object {
        private val kTag = FacebookAdsBridge::class.java.name
        private const val kPrefix = "FacebookAdsBridge"
        private const val kInitialize = "${kPrefix}Initialize";
        private const val kGetTestDeviceHash = "${kPrefix}GetTestDeviceHash"
        private const val kAddTestDevice = "${kPrefix}AddTestDevice"
        private const val kClearTestDevices = "${kPrefix}ClearTestDevices"
        private const val kGetBannerAdSize = "${kPrefix}GetBannerAdSize"
        private const val kCreateBannerAd = "${kPrefix}CreateBannerAd"
        private const val kCreateNativeAd = "${kPrefix}CreateNativeAd"
        private const val kCreateInterstitialAd = "${kPrefix}CreateInterstitialAd"
        private const val kCreateRewardedAd = "${kPrefix}CreateRewardedAd"
        private const val kDestroyAd = "${kPrefix}DestroyAd"
    }

    private var _initializing = false
    private var _initialized = false
    private val _bannerHelper = FacebookBannerHelper()
    private val _ads: MutableMap<String, IAd> = ConcurrentHashMap()

    init {
        _logger.info("$kTag: constructor begin: application = $_application activity = $_activity")
        registerHandlers()
        _logger.info("$kTag: constructor end")
    }

    override fun onCreate(activity: Activity) {
        _activity = activity
        for (ad in _ads.values) {
            ad.onCreate(activity)
        }
    }

    override fun onStart() {}
    override fun onStop() {}

    override fun onResume() {
        for (ad in _ads.values) {
            ad.onResume()
        }
    }

    override fun onPause() {
        for (ad in _ads.values) {
            ad.onPause()
        }
    }

    override fun onDestroy() {
        for (ad in _ads.values) {
            ad.onDestroy()
        }
        _activity = null
    }

    override fun destroy() {
        deregisterHandlers()
        for (ad in _ads.values) {
            ad.destroy()
        }
        _ads.clear()
    }

    @AnyThread
    private fun registerHandlers() {
        _bridge.registerAsyncHandler(kInitialize) {
            Utils.toString(initialize())
        }
        _bridge.registerHandler(kGetTestDeviceHash) {
            testDeviceHash
        }
        _bridge.registerHandler(kAddTestDevice) { message ->
            addTestDevice(message)
            ""
        }
        _bridge.registerHandler(kClearTestDevices) {
            clearTestDevices()
            ""
        }
        _bridge.registerHandler(kGetBannerAdSize) { message ->
            @Serializable
            @Suppress("unused")
            class Response(
                val width: Int,
                val height: Int
            )

            val index = message.toInt()
            val size = _bannerHelper.getSize(index)
            val response = Response(size.x, size.y)
            response.serialize()
        }
        _bridge.registerHandler(kCreateBannerAd) { message ->
            @Serializable
            class Request(
                val adId: String,
                val adSize: Int
            )

            val request = deserialize<Request>(message)
            val adSize = _bannerHelper.getAdSize(request.adSize)
            Utils.toString(createBannerAd(request.adId, adSize))
        }
        _bridge.registerHandler(kCreateNativeAd) { message ->
            @Serializable
            class Request(
                val adId: String,
                val layoutName: String,
                val identifiers: Map<String, String>
            )

            val request = deserialize<Request>(message)
            Utils.toString(createNativeAd(request.adId, request.layoutName, request.identifiers))
        }
        _bridge.registerHandler(kCreateInterstitialAd) { message ->
            Utils.toString(createInterstitialAd(message))
        }
        _bridge.registerHandler(kCreateRewardedAd) { message ->
            Utils.toString(createRewardedAd(message))
        }
        _bridge.registerHandler(kDestroyAd) { message ->
            Utils.toString(destroyAd(message))
        }
    }

    @AnyThread
    private fun deregisterHandlers() {
        _bridge.deregisterHandler(kInitialize)
        _bridge.deregisterHandler(kGetTestDeviceHash)
        _bridge.deregisterHandler(kAddTestDevice)
        _bridge.deregisterHandler(kClearTestDevices)
        _bridge.deregisterHandler(kGetBannerAdSize)
        _bridge.deregisterHandler(kCreateBannerAd)
        _bridge.deregisterHandler(kCreateNativeAd)
        _bridge.deregisterHandler(kCreateInterstitialAd)
        _bridge.deregisterHandler(kCreateRewardedAd)
        _bridge.deregisterHandler(kDestroyAd)
    }

    private fun checkInitialized() {
        if (!_initialized) {
            throw IllegalStateException("Please call initialize() first")
        }
    }

    @AnyThread
    suspend fun initialize(): Boolean {
        return suspendCancellableCoroutine { cont ->
            Thread.runOnMainThread {
                if (AudienceNetworkAds.isInitialized(_application)) {
                    _logger.info("$kTag: initialize: initialized")
                    _initialized = true
                    cont.resume(true)
                    return@runOnMainThread
                }
                if (_initializing) {
                    _logger.info("$kTag: initialize: initializing")
                    cont.resume(false)
                    return@runOnMainThread
                }
                if (_initialized) {
                    _logger.info("$kTag: initialize: initialized")
                    cont.resume(true)
                    return@runOnMainThread
                }
                _initializing = true
                AudienceNetworkAds
                    .buildInitSettings(_application)
                    .withInitListener { result ->
                        if (cont.isActive) {
                            // OK.
                        } else {
                            return@withInitListener
                        }
                        Thread.runOnMainThread {
                            _logger.info("$kTag: initialize: result = ${result.isSuccess} message = ${result.message ?: ""}")
                            _initializing = false
                            _initialized = true
                            if (result.isSuccess) {
                                if (BuildConfig.DEBUG) {
                                    AdSettings.setDebugBuild(true)
                                }
                                AdSettings.setTestMode(false)
                            } else {
                                _logger.error("$kTag: initialize: result = $result")
                            }
                            cont.resume(true)
                        }
                    }
                    .initialize()
            }
        }
    }

    private val testDeviceHash: String
        @AnyThread get() = ""

    @AnyThread
    fun addTestDevice(hash: String) {
        Thread.runOnMainThread {
            checkInitialized()
            AdSettings.addTestDevice(hash)
        }
    }

    @AnyThread
    fun clearTestDevices() {
        Thread.runOnMainThread {
            checkInitialized()
            AdSettings.clearTestDevices()
        }
    }

    @AnyThread
    fun createBannerAd(adId: String, adSize: AdSize): Boolean {
        return createAd(adId) {
            FacebookBannerAd(_bridge, _logger, _activity, adId, adSize, _bannerHelper)
        }
    }

    @AnyThread
    fun createNativeAd(adId: String, layoutName: String,
                       identifiers: Map<String, String>): Boolean {
        return createAd(adId) {
            FacebookNativeAd(_bridge, _logger, _application, _activity, adId, layoutName, identifiers)
        }
    }

    @AnyThread
    private fun createInterstitialAd(adId: String): Boolean {
        return createAd(adId) {
            FacebookInterstitialAd(_bridge, _logger, _activity, adId)
        }
    }

    @AnyThread
    private fun createRewardedAd(adId: String): Boolean {
        return createAd(adId) {
            FacebookRewardedAd(_bridge, _logger, _activity, adId)
        }
    }

    @AnyThread
    fun createAd(adId: String, creator: () -> IAd): Boolean {
        checkInitialized()
        if (_ads.containsKey(adId)) {
            return false
        }
        val ad = creator()
        _ads[adId] = ad
        return true
    }

    @AnyThread
    fun destroyAd(adId: String): Boolean {
        checkInitialized()
        val ad = _ads[adId] ?: return false
        ad.destroy()
        _ads.remove(adId)
        return true
    }
}

File: src/ios/ee/facebook_ads/FacebookAdsBridge.swift
//
//  FacebookAds.swift
//  Pods
//
//  Created by eps on 6/25/20.
//

import FBAudienceNetwork
import RxSwift

private let kTag = "\(FacebookAdsBridge.self)"
private let kPrefix = "FacebookAdsBridge"
private let kInitialize = "\(kPrefix)Initialize"
private let kGetTestDeviceHash = "\(kPrefix)GetTestDeviceHash"
private let kAddTestDevice = "\(kPrefix)AddTestDevice"
private let kClearTestDevices = "\(kPrefix)ClearTestDevices"
private let kGetBannerAdSize = "\(kPrefix)GetBannerAdSize"
private let kCreateBannerAd = "\(kPrefix)CreateBannerAd"
private let kCreateNativeAd = "\(kPrefix)CreateNativeAd"
private let kCreateInterstitialAd = "\(kPrefix)CreateInterstitialAd"
private let kCreateRewardedAd = "\(kPrefix)CreateRewardedAd"
private let kDestroyAd = "\(kPrefix)DestroyAd"

@objc(EEFacebookAdsBridge)
class FacebookAdsBridge: NSObject, IPlugin {
    private let _bridge: IMessageBridge
    private let _logger: ILogger
    private var _initializing = false
    private var _initialized = false
    private let _bannerHelper = FacebookBannerHelper()
    private var _ads: [String: IAd] = [:]

    public required init(_ bridge: IMessageBridge, _ logger: ILogger) {
        _bridge = bridge
        _logger = logger
        super.init()
        registerHandlers()
    }

    public func destroy() {
        deregisterHandlers()
        _ads.values.forEach { $0.destroy() }
        _ads.removeAll()
    }

    func registerHandlers() {
        _bridge.registerAsyncHandler(kInitialize) { _, resolver in
            self.initialize()
                .subscribe(
                    onSuccess: {
                        result in resolver(Utils.toString(result))
                    }, onError: {
                        _ in resolver(Utils.toString(false))
                    })
        }
        _bridge.registerHandler(kGetTestDeviceHash) { _ in
            self.testDeviceHash
        }
        _bridge.registerHandler(kAddTestDevice) { message in
            self.addTestDevice(message)
            return ""
        }
        _bridge.registerHandler(kClearTestDevices) { _ in
            self.clearTestDevice()
            return ""
        }
        _bridge.registerHandler(kGetBannerAdSize) { message in
            let index = Int(message) ?? -1
            let size = self._bannerHelper.getSize(index: index)
            return EEJsonUtils.convertDictionary(toString: [
                "width": size.width,
                "height": size.height
            ])
        }
        _bridge.registerHandler(kCreateBannerAd) { message in
            let dict = EEJsonUtils.convertString(toDictionary: message)
            guard
                let adId = dict["adId"] as? String,
                let index = dict["adSize"] as? Int
            else {
                assert(false, "Invalid argument")
                return ""
            }
            let adSize = self._bannerHelper.getAdSize(index)
            return Utils.toString(self.createBannerAd(adId, adSize))
        }
        _bridge.registerHandler(kCreateNativeAd) { message in
            let dict = EEJsonUtils.convertString(toDictionary: message)
            guard
                let adId = dict["adId"] as? String,
                let layoutName = dict["layoutName"] as? String
            else {
                assert(false, "Invalid argument")
                return ""
            }
            return Utils.toString(self.createNativeAd(adId, layoutName))
        }
        _bridge.registerHandler(kCreateInterstitialAd) { message in
            Utils.toString(self.createInterstitialAd(message))
        }
        _bridge.registerHandler(kCreateRewardedAd) { message in
            Utils.toString(self.createRewardedAd(message))
        }
        _bridge.registerHandler(kDestroyAd) { message in
            Utils.toString(self.destroyAd(message))
        }
    }

    func deregisterHandlers() {
        _bridge.deregisterHandler(kInitialize)
        _bridge.deregisterHandler(kGetTestDeviceHash)
        _bridge.deregisterHandler(kAddTestDevice)
        _bridge.deregisterHandler(kClearTestDevices)
        _bridge.deregisterHandler(kGetBannerAdSize)
        _bridge.deregisterHandler(kCreateBannerAd)
        _bridge.deregisterHandler(kCreateNativeAd)
        _bridge.deregisterHandler(kCreateInterstitialAd)
        _bridge.deregisterHandler(kCreateRewardedAd)
        _bridge.deregisterHandler(kDestroyAd)
    }

    func checkInitialized() {
        if !_initialized {
            assert(false, "Please call initialize() first")
        }
    }

    func initialize() -> Single<Bool> {
        return Single<Bool>.create { single in
            Thread.runOnMainThread {
                if self._initializing {
                    self._logger.info("\(kTag): \(#function): initializing")
                    single(.success(false))
                    return
                }
                if self._initialized {
                    self._logger.info("\(kTag): \(#function): initialized")
                    single(.success(true))
                    return
                }
                self._initializing = true
                FBAudienceNetworkAds.initialize(with: nil) { result in
                    Thread.runOnMainThread {
                        self._logger.debug("\(kTag): initialize: result = \(result.isSuccess) message = \(result.message)")
                        self._initializing = false
                        self._initialized = true
                        if result.isSuccess {
                            FBAdSettings.setAdvertiserTrackingEnabled(true)
                        } else {
                            self._logger.error("\(kTag): initialize: result = \(result.isSuccess)")
                        }
                        single(.success(true))
                    }
                }
            }
            return Disposables.create()
        }
        .subscribeOn(MainScheduler())
    }

    var testDeviceHash: String {
        return FBAdSettings.testDeviceHash()
    }

    func addTestDevice(_ hash: String) {
        Thread.runOnMainThread {
            FBAdSettings.addTestDevice(hash)
        }
    }

    func clearTestDevice() {
        Thread.runOnMainThread {
            self.checkInitialized()
            FBAdSettings.clearTestDevices()
        }
    }

    func createBannerAd(_ adId: String, _ size: FBAdSize) -> Bool {
        return createAd(adId) {
            FacebookBannerAd(_bridge, _logger, adId, size, _bannerHelper)
        }
    }

    func createNativeAd(_ adId: String, _ layoutName: String) -> Bool {
        return createAd(adId) {
            FacebookNativeAd(_bridge, _logger, adId, layoutName)
        }
    }

    func createInterstitialAd(_ adId: String) -> Bool {
        return createAd(adId) {
            FacebookInterstitialAd(self._bridge, self._logger, adId)
        }
    }

    func createRewardedAd(_ adId: String) -> Bool {
        return createAd(adId) {
            FacebookRewardedAd(self._bridge, self._logger, adId)
        }
    }

    func createAd(_ adId: String, _ creator: () -> IAd) -> Bool {
        checkInitialized()
        if _ads.contains(where: { key, _ in key == adId }) {
            return false
        }
        let ad = creator()
        _ads[adId] = ad
        return true
    }

    func destroyAd(_ adId: String) -> Bool {
        checkInitialized()
        guard let ad = _ads[adId] else {
            return false
        }
        ad.destroy()
        _ads.removeValue(forKey: adId)
        return true
    }
}
